File: Java Server/src/messengerserver/Parser.java
package messengerserver;
import java.util.HashMap;



//All messages must contain as their first two letters an Operation Code or Opcode. This Opcode tells the server
//how to parse and handle the incoming message. Specifically, each Opcode is mapped to a bit mask, and a fall-through
//parser is used to parse each message based on this mask. 
//
//Padding of input strings is done using asterisks (*) at this time.
//
//As such, all information in a server message __must__ be in the bitmask appropriate order with the
//appropriate sizes, and a field __must__ not be present if not required by the given Opcode.
//
//The bitmask order is as follows:
//
//000001: Opcode - always 1. 
//000010: UserID - 32 characters - Present for all messages except login and registration. 
//000100: UserName - 32 characters - Present for login, registration, and when adding a friend.
//001000: Password - 128 characters - Present only for login and registration.
//010000: Session ID - 32 characters - Required for all non-login and non-registration interactions. Very weakly verifies connection
//integrity.
//100000: Chat ID - 32 characters - Identifies a single chat between one or multiple people.
//
//The final component of a received transmission, the Message, is whatever remains after the item determined by the bit mask are parsed out.
//
//The core server opcodes with their bitmasks are:
//
//IR (Initial Registration):  001101  /  13
//LR (Login Request):  001101   /   13
//PF (Pull Friends):  011011  /   27
//AF (Add Friend):  010111   / 23
//PC (Pull User-Chat Pairs) / 010011 / 19
//PM (Pull Messages From Chat):  110011    / 51
//SM (Send Message):  110011   /   51

// The core client opcodes with their bitmasks are:
//
// RU (Registration unsuccessful):  000101
// RS (Registration successful):  000101
// LU (Login unsuccessful):	 000101
// LS (Login successful):  010111




public class Parser {
	
	private static final int debugMask = 8; // Indicates the bit mask for Debugger usage. +1 the debugMask to indicate an error message.

	
	// Used to determine the parser behavior.
	
	private static final int userNameLength = Server.MAX_USERNAME_LENGTH;
	private static final int userIDLength = 32;
	private static final int passwordLength = Server.MAX_PASSWORD_LENGTH;
	private static final int sessionIDLength = 32;
	private static final int chatIDLength = 8;
	
	
	
	private static HashMap<String, Integer> opcodeMap;
	
	static {
		
		opcodeMap = new HashMap<String, Integer>();
		
		opcodeMap.put("IR", 13);
		opcodeMap.put("LR", 13);
		opcodeMap.put("PF", 19);
		opcodeMap.put("AF", 23);
		opcodeMap.put("UC", 19);
		opcodeMap.put("PM", 51);
		opcodeMap.put("SM", 51);
	}
	
	
	public static HashMap<String, String> parse(String input) 
	{
		
		HashMap<String, String> returnMap = new HashMap<String, String>();		
		// Gets the opcode.
		String opcode = "ER";
		String message = input;
		
		try {
			opcode = input.substring(0, 2);
			message = input.substring(2);
			returnMap.put("Opcode", opcode);
		}
		catch(Exception e) {
			
			returnMap.put("Opcode", "ER");
			Debugger.record("Parser failed getting opcode", debugMask + 1);
			return returnMap;
		}
		
		
		// Getting the bitmask from the opcode.
		int mask = 0;
		if (opcodeMap.containsKey(opcode)) {
			mask = opcodeMap.get(opcode);

		}
		else {
			
			returnMap.put("Opcode", opcode);
			Debugger.record("Parser failed at bit 0 for opcode: " + opcode + " with input: " + message, debugMask + 1);
			return returnMap;
		}

		// Begin the fall through parser here.
		
		if ((mask & 2) > 0) 
		{
			
			try {
				returnMap.put("UserID", Parser.unpack(message.substring(0, userIDLength)));
				message = message.substring(userIDLength);
				
				Debugger.record("Parser processed at bit 1 for opcode: " + opcode + " with input: " + message, debugMask);

			}
			catch (Exception e) 
			{
				Debugger.record("Parser failed at bit 1 for opcode: " + opcode + " with input: " + message, debugMask + 1);
				returnMap.put("Opcode", "ER");
				return returnMap;
			}
			
		}
		if ((mask & 4) > 0) 
		{

			try {
				Debugger.record("Parser processed at bit 2 for opcode: " + opcode + " with input: " + message, debugMask);

				returnMap.put("UserName", Parser.unpack(message.substring(0, userNameLength)));
				message = message.substring(userNameLength);
				
			}
			catch (Exception e) 
			{
				Debugger.record("Parser failed at bit 2 for opcode: " + opcode + " with input: " + message, debugMask + 1);
				returnMap.put("Opcode", "ER");
				return returnMap;
			}
			
			
		}
		if ((mask & 8) > 0) 
		{

			try {
				Debugger.record("Parser processed at bit 3 for opcode: " + opcode + " with input: " + message, debugMask);

				returnMap.put("Password", Parser.unpack(message.substring(0, 128)));
				message = message.substring(passwordLength);

			}
			catch (Exception e) 
			{
				Debugger.record("Parser failed at bit 3 for opcode: " + opcode + " with input: " + message, debugMask + 1);
				returnMap.put("Opcode", "ER");
				return returnMap;
			}
			
		}
		if ((mask & 16) > 0) 
		{

			try {
				Debugger.record("Parser processed at bit 4 for opcode: " + opcode + " with input: " + message, debugMask);

				returnMap.put("SessionID", message.substring(0, sessionIDLength));
				message = message.substring(sessionIDLength);
			}
			catch (Exception e) 
			{
				Debugger.record("Parser failed at bit 4 for opcode: " + opcode + " with input: " + message, debugMask + 1);
				returnMap.put("Opcode", "ER");
				return returnMap;
			}
			
		}
		if ((mask & 32) > 0) 
		{

			try {
				returnMap.put("ChatID", unpack(message.substring(0, chatIDLength)));
				message = message.substring(chatIDLength);
				
				Debugger.record("Parser processed at bit 5 for opcode: " + opcode + " with input: " + message, debugMask);

			}
			catch (Exception e) 
			{
				Debugger.record("Parser failed at bit 5 for opcode: " + opcode + " with input: " + message, debugMask + 1);
				returnMap.put("Opcode", "ER");
				return returnMap;
			}
			
		}
		
		
		if (message.length() > 0) 
		{
			returnMap.put("Message", message);
		}
		
		return returnMap;
	}
	
	// Used to pack and unpack variable-length components of transmissions such as username or password.
	
	public static String pack(String input, int size) {
		
		String newString = input;
		
        if (newString.length() > size)
        {
        	newString = newString.substring(0, size);
        }

        for (int i = newString.length(); i < size; i++)
        {
        	newString += "*";
        }
        
        Debugger.record("Packed " + input + " into " + newString + "\n", debugMask);

        return newString;
	}

	public static String pack(int input, int size)
	{
		String inputString = String.valueOf(input);
		return pack(inputString, size);
	}
	public static String unpack(String input) {
		
        int packStart = input.indexOf("*");

        String unpackedString = input.substring(0, packStart);

        Debugger.record("Unpacked " + input + " into " + unpackedString + "\n", debugMask);
        
        return unpackedString;
		
	}
}

File: C# Client/Messenger Client/SupportClasses/Parser.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Messenger_Client
{


    /// <summary> 
    /// 
    /// This class contains methods that can be used to parse data from incoming messages.
    /// The primary function returns true if the parse is successful, false otherwise, and
    /// sends output to a provided dictionary. 
    /// 
    /// See MessageDefinitions.txt for details on message structure.
    /// </summary>

    static class Parser
    {

        private static Dictionary<string, int> OpcodeMasks;

        private static int debugMask = 8;


        private const int UserNameLength = 32;
        private const int UserIDLength = 32;
        private const int PasswordLength = 128;
        private const int SessionIDLength = 32;
        private const int ChatIDLength = 8;

        static Parser()
        {
            OpcodeMasks = new Dictionary<string, int>();

            OpcodeMasks.Add("RS", 2);
            OpcodeMasks.Add("RU", 2);
            OpcodeMasks.Add("LU", 0);
            OpcodeMasks.Add("LS", 11);
            OpcodeMasks.Add("LO", 9);
            OpcodeMasks.Add("FP", 9);
            OpcodeMasks.Add("RP", 9);
            OpcodeMasks.Add("UR", 9);
            OpcodeMasks.Add("CP", 9);
            OpcodeMasks.Add("MP", 25);
            OpcodeMasks.Add("CN", 25);
            OpcodeMasks.Add("AM", 9);
            OpcodeMasks.Add("HB", 0);

        }

        /// <param name="input">The input string to the parse function</param>
        /// <param name="outputDict">The dictionary holding the output of the parse function</param>
        public static bool Parse(string input, out Dictionary<string, string> outputDict)
        {

            outputDict = new Dictionary<string, string>();

            string opcode = "";
            string remainder = "";

            try
            {
                opcode = input.Substring(0, 2);
                remainder = input.Substring(2);

                outputDict.Add("Opcode", opcode);
            }
            catch
            {
                Debugger.Record("Parser failed at getting opcode.", debugMask + 1);
                return false;
            }

            // Begin the fall through parser.


            int mask = 0;
            if (OpcodeMasks.ContainsKey(opcode))
            {
                mask = OpcodeMasks[opcode];
            }
            else
            {
                return false;
            }

            // Parse out UserID

            if ((mask & 1) > 0)
            {
                try
                {
                    string userID = Unpack(remainder.Substring(0, UserIDLength));
                    outputDict.Add("UserID", userID);
                    remainder = remainder.Substring(UserIDLength);
                }
                catch
                {
                    Debugger.Record("Parser failed at getting UserID, bit 1.", debugMask + 1);
                    return false;
                }
            }

            // Parse out UserName

            if ((mask & 2) > 0)
            {
                try
                {
                    string username = Unpack(remainder.Substring(0, UserNameLength));
                    outputDict.Add("UserName", username);
                    remainder = remainder.Substring(UserNameLength);
                }
                catch
                {
                    Debugger.Record("Parser failed at getting UserName, bit 2.", debugMask + 1);
                    return false;
                }
            }


            // Parse out Password
            if ((mask & 4) > 0)
            {
                try
                {
                    string password = Unpack(remainder.Substring(0, PasswordLength));
                    outputDict.Add("Password", password);
                    remainder = remainder.Substring(PasswordLength);
                }
                catch
                {
                    Debugger.Record("Parser failed at getting Password, bit 3.", debugMask + 1);
                    return false;
                }
            }

            // Parse out session ID.
            if ((mask & 8) == 8)
            {
                try
                {
                    string sessionID = remainder.Substring(0, SessionIDLength);
                    outputDict.Add("SessionID", sessionID);
                    remainder = remainder.Substring(SessionIDLength);
                }
                catch
                {
                    Debugger.Record("Parser failed at getting SessionID, bit 4.", debugMask + 1);
                    return false;
                }

            }

            if ((mask & 16) == 16)
            {
                try
                {
                    string chatID = Unpack(remainder.Substring(0, ChatIDLength));
                    outputDict.Add("ChatID", chatID);
                    remainder = remainder.Substring(ChatIDLength);
                }
                catch
                {
                    Debugger.Record("Parser failed at getting ChatID, bit 5.", debugMask + 1);
                    return false;
                }
            }

            // Whatever remains is the Message component of the transmission.

            outputDict.Add("Message", remainder);
            return true;
        }



        // This method packs a string in a special null character (currently asterisks) to make it fit the size
        // provided, allowing it to be parsed more easily.
        // Any strings longer than the size given will be truncated.
        /// <param name="input">The string to be packed.</param>
        /// <param name="size">The desired size of the string after packing</param>
        public static string Pack(string input, int size)
        {
            if (input.Length > size)
            {
                input = input.Substring(0, size);
            }

            for (int i = input.Length; i < size; i++)
            {
                input += "*";
            }
            return input;
        }

        /// <summary> 
        /// An overload of the Pack function that allows an integer to be packed.
        /// Relies on the default pack function to operate.
        /// </summary>  
        /// <param name="input">The int to be packed.</param>
		/// <param name="size">The size of the final packed string.</param>
        public static string Pack(int input, int size)
        {
            string inputString = input.ToString();

            return Pack(inputString, size);
        }


        // Unpacks a string that has been packed with filler characters (currently asterisks) to fit the transmission space.
        /// <param name="input">The string to be unpacked.</param>
        private static string Unpack(string packedString)
        {
            int packStart = packedString.IndexOf("*");

            if (packStart < 0)
            {
                return packedString;
            }
            string unpackedString = packedString.Substring(0, packStart);

            return unpackedString;
        }
    }
}
